API_MANIFESTO.md

```diff
@@ -0,0 +1,125 @@
+# The API Manifesto
+
+# Infinispan API dependency module
+
+The ```infinispan-api``` dependency will contain the public API. This module will be independent from the other modules. 
+Dependency with implementations will be **logical**. This will keep the modularity needed for the different use cases.
+
+Examples:
+ * Cache use case
+ 
+ * NoSQL data store use case
+ 
+ * Tasks, concurrency, counters uses cases
+
+The dependencies to thirty-party libraries must be avoided. This module should contain dependencies to JDK classes, 
+not more.
+
+If you really **need** for a reason to use an external library to be part of an API. 
+Here is a non-exhaustive list of what we should discuss before promoting an API that uses an external dependency:
+
+-Is there a way to keep that library in implementation instead of public API? Creating our Infinispan 
+public classes and keeping it internal
+- Are we are being too opinionated?
+- Who are we depending on? What is going to happen when this library won't be maintained or will break?
+- Is this library such important that is considered to be included in the JDK in the near future?
+
+**The promotion to public official API should be done by care and debate. 
+By default we don't use other classes than JDK + ours.**
+
+# User First
+
+* Before making an API public, consider to test the usability with someone not familiar with the feature
+
+* Convention over Configuration: Make sure a default behaviour exists. Don't make me think.
+
+* Avoid exposing internals as much as possible. Even if your feature lays on top of a cache today, internals can change.
+
+* Each public API/Feature must have a simple tutorial showcasing the example. Without a demo, the API does not exist
+
+* Focus on use cases over features. Try to showcase in a demo, blog post and tutorial how to solve a problem they have
+
+* Ask for user's feedback actively (zulip, user forums, blog post, twitter, github ...)
+
+* Be clear about whats PUBLIC API and what internals/implementations are. Place your implementations under `impl.*` package.
+
+# Documentation
+
+* Public API must have Java Doc
+
+* Public API must have User Doc 
+  
+* Public API must have a Simple Tutorial (written + code)   
+
+* Public API must have at least 1 blog post
```

Comment:
In Vert.x we're doing all of the above except blog post. The downside of blog post is that they get old. We try to favor posts linking to howtos now.

Reply:
Ok, let's do that